src/common/dav_v2/heap.c

```diff
@@ -955,36 +978,161 @@ heap_reclaim_zone_garbage(struct palloc_heap *heap, struct bucket *bucket,
 		return -1;
 
-	i = h->zones_ne_gc ? h->zones_lastne_gc + 1 : 0;
-
-	for (; i < h->zones_exhausted; i++) {
+	for (i = h->zones_nextne_gc; i < h->zones_exhausted; i++) {
 		heap_zinfo_get(heap, i, &allotted, &evictable);
 		if (!allotted)
-			break;
+			continue;
 		if (!evictable) {
+			h->zones_nextne_gc = i + 1;
 			*zone_id = i;
 			return 0;
 		}
 	}
 	return -1;
 }
 
+static void
+heap_reclaim_setlast_ne(struct palloc_heap *heap, uint32_t zone_id)
+{
+	if (zone_id >= heap->rt->zones_nextne_gc)
+		heap->rt->zones_nextne_gc = zone_id + 1;
+}
+
+static int
+heap_get_next_unused_zone(struct palloc_heap *heap, uint32_t *zone_id)
+{
+	bool     allotted, evictable;
+	uint32_t i;
+
+	for (i = heap->rt->zones_unused_first; i < heap->rt->nzones; i++) {
+		heap_zinfo_get(heap, i, &allotted, &evictable);
+		if (!allotted)
+			break;
+	}
+	if (i == heap->rt->nzones) {
+		heap->rt->zones_unused_first = heap->rt->nzones;
+		return -1;
+	}
+
+	*zone_id = i;
+	return 0;
+}
+
+static int
+heap_mark_zone_used_transient(struct palloc_heap *heap, uint32_t zone_id, bool is_evictable)
+{
+	struct mbrt *mb;
+
+	if (is_evictable) {
+		mb = heap_mbrt_setup_mb(heap, zone_id);
+		if (mb == NULL) {
+			ERR("Failed to setup mbrt for zone %u\n", zone_id);
+			return -1;
+		}
+		heap_mbrt_setmb_evictable(heap, mb);
+	} else
+		heap_mbrt_setmb_nonevictable(heap, zone_id);
+
+	heap->rt->zones_unused_first = zone_id + 1;
+	if (heap->rt->zones_exhausted < heap->rt->zones_unused_first)
+		heap->rt->zones_exhausted = heap->rt->zones_unused_first;
+	if (is_evictable)
+		heap->rt->zones_exhausted_e++;
+	else
+		heap->rt->zones_exhausted_ne++;
+	return 0;
+}
+
+static void
+heap_mark_zone_used_persist(struct palloc_heap *heap, uint32_t zone_id)
+{
+	bool is_evictable = heap_mbrt_ismb_evictable(heap, zone_id);
+
+	if (zone_id)
+		heap_zinfo_set(heap, zone_id, true, is_evictable);
+}
+
+static void
+heap_mark_zone_unused_transient(struct palloc_heap *heap, uint32_t zone_id)
+{
+	struct mbrt *mb = heap_mbrt_get_mb(heap, zone_id);
+
+	if (heap_mbrt_ismb_evictable(heap, zone_id)) {
+		D_ASSERT(mb != NULL);
+		heap_mbrt_cleanup_mb(mb);
+		heap->rt->zones_exhausted_e--;
+	} else
+		heap->rt->zones_exhausted_ne--;
+
+	heap_mbrt_setmb_unused(heap, zone_id);
+
+	if (heap->rt->zones_unused_first > zone_id)
+		heap->rt->zones_unused_first = zone_id;
+	if (heap->rt->zones_exhausted == (zone_id + 1))
+		heap->rt->zones_exhausted = zone_id;
+}
+
+static void
+heap_mark_zone_unused(struct palloc_heap *heap, uint32_t zone_id)
+{
+	struct umem_cache_range rg           = {0};
+	bool                    is_evictable = heap_mbrt_ismb_evictable(heap, zone_id);
+	int                     rc;
+
+	D_ASSERT(is_evictable == false);
+
+	heap_mark_zone_unused_transient(heap, zone_id);
+	rg.cr_off = GET_ZONE_OFFSET(zone_id);
+	rg.cr_size =
+	    ((heap->size - rg.cr_off) > ZONE_MAX_SIZE) ? ZONE_MAX_SIZE : heap->size - rg.cr_off;
+	rc = umem_cache_map(heap->layout_info.store, &rg, 1);
```

Comment:
hmm, will it cause the page being loaded if the zone is E? That'll cause problem since it could be in a tx.